token-swap/program/src/constraints.rs

```diff
@@ -0,0 +1,184 @@
+//! Various constraints as required for production environments
+
+use crate::{
+    curve::{SwapCurve, CurveType, ConstantProductCurve, FlatCurve},
+    error::SwapError,
+};
+
+use solana_program::program_error::ProgramError;
+
+#[cfg(feature = "production")]
+use std::env;
+
+/// Encodes fee constraints, used in multihost environments where the program
+/// may be used by multiple frontends, to ensure that proper fees are being
+/// assessed.
+/// Since this struct needs to be created at compile-time, we only have access
+/// to const functions and constructors. Since SwapCurve contains a Box, it
+/// cannot be used, so we have to split the curves based on their types.
+pub struct FeeConstraints<'a> {
+    /// Owner of the program
+    pub owner_key: &'a str,
+    /// Valid flat curves for the program
+    pub valid_flat_curves: &'a[FlatCurve],
+    /// Valid constant product curves for the program
+    pub valid_constant_product_curves: &'a[ConstantProductCurve],
```

Comment:
I'm not super happy about having to split these up, but we can't really have `dyn` things in const structure